File: zadanie.cs
using System;

class Program
{
    static void Main()
    {
        string str = "sd10sds101sdd100101"; // строка
        int inState = 0;
        string output = "";

        for (int i = 0; i < str.Length; i++)
        {
            if (str[i] == '1' && inState == 0)
            {
                inState = 1;
                output = "1";
                continue;
            }
            if (str[i] == '1' && inState == 2)
            {
                output += "1";
                Console.WriteLine(output);
                inState = 0;
                i--;
            }
            if (str[i] == '0' && (inState == 1 || inState == 2))
            {
                output += "0";
                inState = 2;
            }
            else if (inState == 1 || inState == 2)
            {
                inState = 0;
            }
        }
    }
}

File: zadanie.js
let str = "sd10sds101sdd100101"; // строка
let inVar = 0;
let out = "";

for (let i = 0; i < str.length; i++) {
    if (str[i] === '1' && inVar === 0) {
        inVar = 1;
        out = '1';
        continue;
    }
    if (str[i] === '1' && inVar === 2) {
        out += '1';
        console.log(out);
        inVar = 0;
        i--;
    }
    if (str[i] === '0' && (inVar === 1 || inVar === 2)) {
        out += '0';
        inVar = 2;
    } else if (inVar === 1 || inVar === 2) {
        inVar = 0;
    }
}
